### ISSUE TITLE
[java] UselessPureMethodCall: False positive for Stream.forEach

### ISSUE THREAD

**Issue by estekhin** [pmd/pmd#6781]
**Affects PMD Version:** 7.25.0

**Rule:** https://docs.pmd-code.org/pmd-doc-7.25.0/pmd_rules_java_errorprone.html#uselesspuremethodcall

**Description:**

PMD 7.25.0 reports *some* usages of `Stream.forEach` when forEach is used to add elements to an existing collection declared as `var`. The same code works ok when that collection type is declared explicitly.

The only difference between methods in the sample code is the declaration of the `result` variable.

PMD 7.24.0 works fine.

**Code Sample demonstrating the issue:**

```java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class UselessPureMethodCallStreamForEach {

    List<Object> ok(List<Object> source) {
        ArrayList<Object> result = source.stream()
            .collect(Collectors.toCollection(ArrayList::new));

        source.stream()
            .forEach(result::add);

        return result;
    }

    List<Object> triggersUselessPureMethodCall(List<Object> source) {
        var result = source.stream()
            .collect(Collectors.toCollection(ArrayList::new));

        // the line below triggers
        // UselessPureMethodCallStreamForEach:25 Rule:UselessPureMethodCall Priority:3 Do not call pure method forEach if the result is not used..
        source.stream()
            .forEach(result::add);

        return result;
    }

}
```

**Expected outcome:**

PMD should not report UselessPureMethodCall for this code.

**Running PMD through:** *Maven*

**Comment by UncleOwen** [comment]
A side-effect of the changes to `InvocationMatcher` in #6712.

In [JavaRuleUtil.isKnownPure()](https://github.com/pmd/pmd/blob/7753d9e3e097624d5ddbc2bed95a4c1237019213/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/JavaRuleUtil.java#L458), `KNOWN_PURE_METHODS.anyMatch(call)` now returns true (since the argument isn't resolved) and `!call.getMethodType().getReturnType().isVoid()` also returns true (also because the argument isn't resolved). @zbynek I don't think that was intended?
